icrestimator.py

```diff
@@ -286,25 +320,48 @@ def S(self, lmda: np.ndarray):
         :return: row vector expressing the point.
         """
         S = np.zeros(shape=(self.n_modules,))
-        lmda = lmda.T # computations require lambda as a row vector
+        lmda = lmda.T  # computations require lambda as a row vector
         for i in range(self.n_modules):
             # equations 16 and 17 in the paper
             a = column(self.a, i)
             a_orth = column(self.a_orth, i)
             l = column(self.l_v, i)
-            delta = lmda.dot(a-l)
+            delta = lmda.dot(a - l)
             omega = lmda.dot(a_orth)
             norm = np.linalg.norm([delta, omega])
             sin_beta = np.sign(delta) * omega / norm
             cos_beta = np.sign(delta) * delta / norm
             S[i] = math.atan2(sin_beta, cos_beta)
         return S
 
+    def flip_wheel(self, q: np.ndarray, S_lmda: np.ndarray):
+        """
+        Determine if the wheel is either already facing the desired direction or is out by pi,
+        in both cases the wheel does not have to turn.
+        :param q: an array representing all of the current beta angles
+        :parem S_lmda: an array of all the beta angles required to achieve a desired ICR
+        :return: an array of the same length as the input arrays with each component
+        as the correct distance of q from S_lmda.
+        This is done to prevent an 'out by pi' issue where q and S_lmda would not converge.
+        We also track the number of times each wheel has been flipped to ensure that it drives
+        in the correct direction, True indicates drive direction should be reversed.
+        """
+        output = np.array([None] * self.n_modules)
+        for module in range(self.n_modules):
+            dif = q[module] - S_lmda[module]
+            output[module] = math.atan(math.sin(dif) / math.cos(dif))
+            absolute_direction = math.atan2(math.sin(dif), math.cos(dif))
+            if abs(output[module] - absolute_direction) > self.tolerance:
+                self.flipped[module] = True
+            else:
+                self.flipped[module] = False
+        return output
```

Comment:
It seems you could rather easily vectorise this code rather than writing a for loop. Mind having a go at that?